File: GolfIt/Vector.cs
namespace GolfIt
{
    public class Vector
    {
        public float X, Y;

        public Vector(float x, float y)
        {
            this.Y = (float)y;
            this.X = (float)x;

        }


        public static Vector operator -(Vector v)
        {
            return new Vector(-v.X, -v.Y);
        }


        public static Vector operator +(Vector a, Vector b)
        {
            return new Vector(a.X + b.X, a.Y + b.Y);
        }

        public static Vector operator -(Vector a, Vector b)
        {
            return new Vector(a.X - b.X, a.Y - b.Y);
        }

        public static Vector operator *(float a, Vector v)
        {
            return new Vector(a * v.X, a * v.Y);
        }

        public static Vector operator *(Vector v, float a)
        {
            return new Vector(a * v.X, a * v.Y);
        }

        public static Vector operator /(Vector v, float a)
        {
            return new Vector(v.X / a, v.Y / a);
        }

        public float MagSQR()
        {
            float f = (X * X) + (Y * Y);
            return f;
        }

        public float Length()
        {
            float f = (float)Math.Sqrt((X * X) + (Y * Y));
            return f;
        }


        public float Distance(Vector a)
        {
            float f = (float)Math.Sqrt(Math.Pow(X - a.X, 2) + Math.Pow(Y - a.Y, 2));
            return f;
        }

        public Vector Normalized()
        {
            float magnitude = Length();
            if (magnitude == 0) throw new InvalidOperationException("Cannot normalize a zero vector.");
            return this / magnitude;
        }
    }
}

File: golfit_rs/src/vector.rs
#[derive(Debug, Clone, Copy)]
pub struct Vector2D {
    pub x: f32,
    pub y: f32,
}

impl Vector2D {
    pub fn new(x: f32, y: f32) -> Self {
        Self { x, y }
    }

    pub fn zero() -> Self {
        Self { x: 0.0, y: 0.0 }
    }

    pub fn length(&self) -> f32 {
        (self.x * self.x + self.y * self.y).sqrt()
    }

    pub fn mag_sqr(&self) -> f32 {
        self.x * self.x + self.y * self.y
    }

    pub fn distance(&self, other: &Vector2D) -> f32 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }

    pub fn normalized(&self) -> Vector2D {
        let magnitude = self.length();
        if magnitude == 0.0 {
            Vector2D::zero()
        } else {
            *self / magnitude
        }
    }

    pub fn dot(&self, other: &Vector2D) -> f32 {
        self.x * other.x + self.y * other.y
    }

    pub fn cross(&self, other: &Vector2D) -> f32 {
        self.x * other.y - self.y * other.x
    }
}

impl std::ops::Add for Vector2D {
    type Output = Vector2D;

    fn add(self, other: Vector2D) -> Vector2D {
        Vector2D {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }
}

impl std::ops::Sub for Vector2D {
    type Output = Vector2D;

    fn sub(self, other: Vector2D) -> Vector2D {
        Vector2D {
            x: self.x - other.x,
            y: self.y - other.y,
        }
    }
}

impl std::ops::Mul<f32> for Vector2D {
    type Output = Vector2D;

    fn mul(self, scalar: f32) -> Vector2D {
        Vector2D {
            x: self.x * scalar,
            y: self.y * scalar,
        }
    }
}

impl std::ops::Mul<Vector2D> for f32 {
    type Output = Vector2D;

    fn mul(self, vector: Vector2D) -> Vector2D {
        Vector2D {
            x: self * vector.x,
            y: self * vector.y,
        }
    }
}

impl std::ops::Mul<Vector2D> for Vector2D {
    type Output = Vector2D;

    fn mul(self, other: Vector2D) -> Vector2D {
        Vector2D {
            x: self.x * other.x,
            y: self.y * other.y,
        }
    }
}

impl std::ops::Div<f32> for Vector2D {
    type Output = Vector2D;

    fn div(self, scalar: f32) -> Vector2D {
        Vector2D {
            x: self.x / scalar,
            y: self.y / scalar,
        }
    }
}

impl std::ops::Neg for Vector2D {
    type Output = Vector2D;

    fn neg(self) -> Vector2D {
        Vector2D {
            x: -self.x,
            y: -self.y,
        }
    }
}
